src/openassetio-python/tests/bridge/python/test_converter.cpp

```diff
@@ -0,0 +1,268 @@
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+
+  GIVEN("A Python object casted to a C++ object") {
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    const TraitsDataPtr traitsData =
+        openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+    WHEN("data is set via the C++ object") {
+      traitsData->addTrait(kTestTraitId);
+
+      THEN("Python object reflects that data set") {
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+    }
+
+    WHEN("data is set via the Python object") {
+      PyObject_CallMethod(pyTraitsData, "addTrait", "s", kTestTraitId);
+      THEN("C++ object reflects the data set") { CHECK(traitsData->hasTrait(kTestTraitId)); }
+    }
+    Py_DECREF(pyTraitsData);
+  }
+}
+
+SCENARIO("Casting to PyObject extends object lifetime") {
+  GIVEN("A Python object casted from a C++ object") {
+    TraitsDataPtr traitsData = TraitsData::make();
+    traitsData->addTrait(kTestTraitId);
+    PyObject* pyTraitsData = openassetio::python::converter::castToPyObject(traitsData);
+
+    WHEN("C++ reference is destroyed") {
+      CHECK(Py_REFCNT(pyTraitsData) == 1);  // Initial condition.
+      traitsData.reset();
+
+      THEN("object remains alive and can be operated on via the Python interpreter") {
+        CHECK(Py_REFCNT(pyTraitsData) == 1);
+        PyObject* resultBool = PyObject_CallMethod(pyTraitsData, "hasTrait", "s", kTestTraitId);
+
+        // Use CHECK rather than REQUIRE in order to ensure the Decrefs
+        // below actually execute.
+        CHECK(PyBool_Check(resultBool));
+        CHECK(PyObject_IsTrue(resultBool));
+        Py_DECREF(resultBool);
+      }
+      Py_DECREF(pyTraitsData);
+    }
+  }
+}
+
+SCENARIO("Casting to a C++ object binds object lifetime") {
+  GIVEN("A python object") {
+    py::module_::import("openassetio");
+
+    // Create a python object and release it, to simulate an unmanaged
+    // PyObject being provided to us.
```

Comment:
[fix] We don't make the same qualifying comment about L69 - is this relevant still? Is it that we're making a _pybind_ object that motivates the callout to explain the `release()`  as we'd just `PyObject*` natively if it wasn't cumersome? if so, can we say pybind and it's convenient, and add the comment before too (+ update 150)?  I understood that we wanted to use raw CPython where possible in these tests, so good to have the exception explained (and maybe that philosophy documented at the top too?)